### PR TITLE
Add tag support to LogGroup resource

### PR DESCRIPTION
**Pull request by gruebel**
Fixes aws-cloudformation/aws-cloudformation-coverage-roadmap#77

*Description of changes:*
I added the possibility to add tags to the LogGroup resource. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by wbingli** [comment]
Thanks for your contribution. We are prioritizing this PR review.

**Comment by petewilcock** [comment]
+1 on this PR, we're awaiting this functionality 💯

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java
Comment (PatMyron) [inline aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:36]: we probably shouldn't maintain this list in multiple places
```suggestion
```
https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-logs/blob/0406d2bd4e31329bfb2b8a10b97a21cebeaf4d79/aws-logs-loggroup/aws-logs-loggroup.json#L121-L123

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java
Comment (PatMyron) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:190]: https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/pull/306 should handle `AwsServiceExceptions` automatically but [the contract seems to still require some to be explicitly mapped](https://github.com/aws-cloudformation/cloudformation-cli/issues/527)

Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:189]: I added one more exception handling, because the link provided is actually the wrong one, it is for CloudWatch not CloudWatch Logs. That's the correct one 😄 
https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_TagLogGroup.html#API_TagLogGroup_Errors

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Translator.java
Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Translator.java:165]: if I don't return `null` here some of the contract tests will fail with
```
{'status': 'FAILED', 'errorCode': 'GeneralServiceException', 'message': "1 validation error detected: Value '{}' at 'tags' failed to satisfy constraint: Member must have length greater than or equal to 1 (Service: CloudWatchLogs, Status Code: 400, Request ID: 710f70c1-0dad-4f89-91ad-3eadd83b6e9c, Extended Request ID: null)", 'callbackDelaySeconds': 0}
```

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Translator.java
Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Translator.java:164]: will result in same error as below

File: aws-logs-loggroup/pom.xml
Comment (PatMyron) [inline aws-logs-loggroup/pom.xml:46]: There have been issues with resource providers using newer AWS SDK versions than the Java plugin is using:
https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/pull/267#issuecomment-632796555
https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/pull/324

Is the current AWS SDK version the Java plugin is using sufficient?
```suggestion
            <version>2.15.19</version>
```

Comment (gruebel) [reply aws-logs-loggroup/pom.xml]: sure, changed it and moved the version to the properties section

**Comment by gruebel** [comment]
@PatMyron is there anything else I should change or more feedback?

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java
Comment (wbingli) [inline aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java:33]: To have backward compatibility, the ListTagsLogGroups call should be soft-fail if no permission for tagging.

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java
Comment (wbingli) [inline aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:40]: Should use `getDesiredResourceTags` which will combine both stack tags and resource tag.  And `getPreviousResourceTags` for both previous stack tags and resource tags. 

You will need to override the `resourceDefinedTags` ([example](https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-kms/blob/master/key/src/main/java/software/amazon/kms/key/Configuration.java#L20)) so that the `getDesiredResourceTags` will merge the stacks and resource defined tags together.

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java
Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java:33]: sure, I can add it, but I don't get why I should support here backward compatibility? Shouldn't I then do it everywhere, where I interact with tags in some way? I'm just really curious about it, hope to get some more insights 😄

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java
Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:40]: Good point, makes sense and will try to add it.

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java
Comment (PatMyron) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java:33]: particularly an issue with `ReadHandler` since permissions are expanding without customer changes

Comment (wbingli) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java:33]: @gruebel  Some users may won't have tagging permissions, especially they use stack role and restrict the permissions to operate a stack . 

In this way, it will break people if this change go out.   It makes senses to fail if users specify tags during create or update time. However, if they creates the log group successfully before,  it will be bad experience to break them.  BTW, the READ will be used for `GetAtt` function to get the attribute data.

Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java:33]: Thanks for the insight. Makes sense with the restocked permission for stack updates. I also used it in my previous work.

Comment (gruebel) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java:33]: What I'm still curious about is, what is actually the List handler used for. The Read one I understand and it is used also during stack resource import, but the List one?!

**Comment by gruebel** [comment]
@wbingli would be great to get some feedback for my new changes 😄

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java
Comment (wbingli) [reply aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/ReadHandler.java]: This List handler is not used in CloudFormation yet, it intends for future features, so having it hard fail for permission issue is right to do.

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java
Comment (wbingli) [inline aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/UpdateHandler.java:202]: We need to soft fail the Tag permission as well for Update.  The reason is that the Tagging change would comes from stack tags as well. The use case is that some customers don't give Tag permission to the stack.  Before this change, the Log group resource can be update successful while doing stack tag update, as this resource just ignore any tag update.  

With this change, it will now update the tag for cloudformation stack tags and it can run into permission error and fails the resource update.  Soft fail with it as same as read handler can avoid breaking change to those customers.

**Review by wbingli — CHANGES_REQUESTED** [review]

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/CreateHandler.java
Comment (wbingli) [inline aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/CreateHandler.java:29]: Nice. Tag on create is the right way to create tags in a following API call. It can allow customer to control the resource permission with tags.

File: aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Configuration.java
Comment (wbingli) [inline aws-logs-loggroup/src/main/java/software/amazon/logs/loggroup/Configuration.java:27]: Nice to resolve the conflict to merge the tags into map.  The tags could have duplication and without the resolving duplication it can run into error.

**Review by wbingli — CHANGES_REQUESTED** [review]

**Comment by jaglade** [comment]
Is there any update on this PR? We are eagerly awaiting this functionality.

**Comment by wangyun1517** [comment]
Any update on this PR?

**Comment by Aleksandr-Filichkin** [comment]
Any update on this?

**Comment by vitali-pozniak** [comment]
+1

**Comment by mgmarino** [comment]
It'd be great to have input from either the reviewers or @gruebel (who has already put in a lot of work here, thanks!) about whether help from the community is desired to get this over the finish line.  This is a very important feature improvement.

**Comment by JoryUK** [comment]
+1

**Comment by gruebel** [comment]
sorry, everyone. I totally forgot about my PR 🙈 

@wbingli I added the silent fail behaviour for add and remove tags, when trying to update a log group. Tagging permissions are not needed when creating a log group with tags.

**Review by wbingli — APPROVED** [review]

**Comment by wbingli** [comment]
@gruebel  Thanks for your contribution.  Will merge it once we have other reviewer approve it.

**Review by ammokhov — APPROVED** [review]

**Comment by shoddyknight** [comment]
I am eagerly awaitiing this being merged so I can tag and find my cfn-init logs easily

**Comment by samjarrett** [comment]
Is this still yet to be deployed? It definitely isn't [documented](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-logs-loggroup.html) and doesn't work if you specify `Tags:` as a property...

**Comment by sfbaker7** [comment]
Hi, we are expecting this to be available in all regions by the end of next week. Documentation change will follow after

Will update this thread once deployed

**Comment by glc-froussel** [comment]
Hello there, any update on this ?

**Comment by glc-froussel** [comment]
Hello there, any update on this ?

**Comment by bool3012** [comment]
Any update?? This feature still doesn't work, at least in eu regions..

**Comment by felixfang** [comment]
Hi this is Felix from AWS, to add support of tags, our team need to go through security review to release it. We are currently working on the review.

**Comment by bencovi** [comment]
Hi @felixfang - any updates?

**Comment by felixfang** [comment]
Hi we are still working on the security review. We are aiming to have it ready on September

**Comment by Karthi96** [comment]
@felixfang  any update on this, when this can be released?

**Comment by reply2srij** [comment]
Can we get an update on this please?

**Comment by glc-froussel** [comment]
> Can we get an update on this please?

It has been released on November 22 ;)

**Comment by SeijiSuenaga** [comment]
Is it just me, or are stack-level tags not always applied to log group resources in the stack?  I'm seeing inconsistent results where log groups don't always (but sometimes do) have all the tags that are on the stack.

**Comment by glc-froussel** [comment]
> Is it just me, or are stack-level tags not always applied to log group resources in the stack? I'm seeing inconsistent results where log groups don't always (but sometimes do) have all the tags that are on the stack.

I had the same behaviour and the cause was that I had to explicitly create the log group resource in the CF template instead of having the related resource create it (aka, the log group must be part of your stack resources).

Maybe you are facing the same issue ?

**Comment by mgmarino** [comment]
> Is it just me, or are stack-level tags not always applied to log group resources in the stack? I'm seeing inconsistent results where log groups don't always (but sometimes do) have all the tags that are on the stack.

CF will only propagate *changes*.  In other words, if you have stack-level tags and they have not changed since this feature was introduced, they will not get propagated to the log groups.  For already existing stacks, this was an issue for us since we have some tags that do not regularly change on deployments.  (Some, like the Version, do, and these were getting propagated as expected.)  To force the tags to get propagated, we had to do some manual deployments setting the tags to a temporary value (note, just omitting them will not "delete" them).

**Comment by SeijiSuenaga** [comment]
> > Is it just me, or are stack-level tags not always applied to log group resources in the stack? I'm seeing inconsistent results where log groups don't always (but sometimes do) have all the tags that are on the stack.
> 
> CF will only propagate _changes_. In other words, if you have stack-level tags and they have not changed since this feature was introduced, they will not get propagated to the log groups. For already existing stacks, this was an issue for us since we have some tags that do not regularly change on deployments. (Some, like the Version, do, and these were getting propagated as expected.) To force the tags to get propagated, we had to do some manual deployments setting the tags to a temporary value (note, just omitting them will not "delete" them), and then a subsequent deployment with the correct tag values.

Thank you, that's exactly what's happening in my case, so I'll have to do some manual shenanigans too unfortunately.

**Pull request merged**
